### PR TITLE
bugfix(parser): allow multiple or conditions in `or` and `and` clauses

### PR DESCRIPTION
**Pull request by volokluev**
### Context

Even though `or(1, 1, 1, 1)` is valid snql, the clickhouse formatter chokes on it because there's an assumption in [other parts of the code](https://github.com/getsentry/snuba/blob/master/snuba/query/conditions.py#L237-L243) which assumes that ands and ors have two arguments. Breaking that assumption causes an infinite recursion error

### Solution

Rather than changing the matching infrastructure and making sure all the query processors using that assumption work afterwards, I just massaged the AST post-parsing to make sure that all `or` and `and` functions had at most two arguments

### CODE REVIEW

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -1478,12 +1513,16 @@ def parse_snql_query(
...
 ) -> Tuple[Union[CompositeQuery[QueryEntity], LogicalQuery], str]:
+
     with sentry_sdk.start_span(op="parser", description="parse_snql_query_initial"):
         query = parse_snql_query_initial(body)
 
     if settings and settings.get_dry_run():
         explain_meta.set_original_ast(str(query))
 
+    with sentry_sdk.start_span(op="processor", description="treeify_conditions"):
+        _post_process(query, [_treeify_or_and_conditions], settings)
```
Comment (enochtangg) [inline snuba/query/snql/parser.py:1524]: Looks good! Quick question, is there a reason why we shouldn't just shove this into `POST_PROCESSORS`?

Comment (volokluev) [reply snuba/query/snql/parser.py:1527]: Good question, explained in comment

File: snuba/query/snql/parser.py [hunk 2]
```diff
@@ -988,6 +988,40 @@ def transform(exp: Expression) -> Expression:
...
+    Adding this post-process step is easier than changing the rest of the query pipeline
+    """
+
+    def transform(exp: Expression) -> Expression:
+        if not isinstance(exp, FunctionCall):
+            return exp
+
+        if exp.function_name not in ("and", "or"):
+            return exp
+        args = exp.parameters
```
Comment (evanh) [inline snuba/query/snql/parser.py:1010]: There are helper functions that will do this for you already: https://github.com/getsentry/snuba/blob/master/snuba/query/conditions.py#L278

File: snuba/query/snql/parser.py [hunk 3]
```diff
@@ -1478,12 +1513,19 @@ def parse_snql_query(
...
     settings: QuerySettings | None = None,
 ) -> Tuple[Union[CompositeQuery[QueryEntity], LogicalQuery], str]:
+
     with sentry_sdk.start_span(op="parser", description="parse_snql_query_initial"):
         query = parse_snql_query_initial(body)
 
     if settings and settings.get_dry_run():
         explain_meta.set_original_ast(str(query))
 
+    # NOTE (volo): The anonimizer that runs after this function call chokes on
```
Comment (evanh) [inline snuba/query/snql/parser.py:1523]:
```suggestion
    # NOTE (volo): The anonymizer that runs after this function call chokes on
```

File: snuba/query/snql/parser.py [hunk 4]
```diff
@@ -1478,12 +1513,19 @@ def parse_snql_query(
...
+    # OR and AND clauses with multiple parameters so we have to teeify them
```
Comment (evanh) [inline snuba/query/snql/parser.py:1524]:
```suggestion
    # OR and AND clauses with multiple parameters so we have to treeify them
```

**Review by enochtangg — APPROVED** [review]

**Pull request merged**